guides/src/diagnostics/ruby.md

```diff
@@ -0,0 +1,1438 @@
+# Diagnostics for Ruby
+Each error description will follow the format below.
+
+````
+## Error Name
+
+**Description**: Brief error description
+
+**Example:**
+
+```ruby
+Ruby code that detects the error
+```
+
+```
+Error messages obtained by running `steep check`
+```
+
+**severity**:
+
+A table showing how the severity of the corresponding error is set in Steep's error presets
+````
+
+
+## Ruby::ArgumentTypeMismatch
+
+**Description**: Occurs when the method types do not match.
+
+**Example**:
+
+```ruby
+'1' + 1
+```
+
+```
+test.rb:1:6: [error] Cannot pass a value of type `::Integer` as an argument of type `::string`
+│   ::Integer <: ::string
+│     ::Integer <: (::String | ::_ToStr)
+│       ::Integer <: ::String
+│         ::Numeric <: ::String
+│           ::Object <: ::String
+│             ::BasicObject <: ::String
+│
+│ Diagnostic ID: Ruby::ArgumentTypeMismatch
+│
+└ '1' + 1
+        ~
+```
+
+**severity**:
+
+| all_error | default | strict | lenient | silent |
+| --- | --- | --- | --- | --- |
+| error | error | error | information | nil |
+
+## Ruby::BlockBodyTypeMismatch
+
+**Description**: Occurs when the return type of the block body does not match the expected type.
+
+**Example**:
+
+```ruby
+lambda {|x| x + 1 } #: ^(Integer) -> String
+```
+
+```
+test.rb:1:7: [error] Cannot allow block body have type `::Integer` because declared as type `::String`
+│   ::Integer <: ::String
+│     ::Numeric <: ::String
+│       ::Object <: ::String
+│         ::BasicObject <: ::String
+│
+│ Diagnostic ID: Ruby::BlockBodyTypeMismatch
+│
+└ lambda {|x| x + 1 } #: ^(Integer) -> String
+         ~~~~~~~~~~~~
+```
+
+**severity**:
+
+| all_error | default | strict | lenient | silent |
+| --- | --- | --- | --- | --- |
+| error | warning | error | information | nil |
```

Comment:
Looks like `default` is between `strict` and `lenient`.
```suggestion
| all_error | strict | default | lenient | silent |
| --- | --- | --- | --- | --- |
| error | error | warning | information | nil |
```
Also, I’m not sure if it’s necessary to include the `all_error` and `silent` severities, as they (should) set everything to `error` and `nil` respectively.

Reply:
> Looks like `default` is between `strict` and `lenient`.

Thanks! I'll fix it.

> Also, I’m not sure if it’s necessary to include the `all_error` and `silent` severities, as they (should) set everything to `error` and `nil` respectively.

This was a point that bothered me too. I thought it would be valuable to have a list of error presets, so I'm listing them here, but I'd like to hear soutaro-san's thoughts.